src/main/java/mate/academy/springlibrary/service/BookServiceImpl.java

```diff
@@ -18,12 +20,15 @@
 @RequiredArgsConstructor
 public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
+    private final CategoryRepository categoryRepository;
     private final BookMapper bookMapper;
     private final BookSpecificationBuilder bookSpecificationBuilder;
 
     @Override
     public BookDto save(CreateBookRequestDto requestDto) {
         Book book = bookMapper.toModel(requestDto);
+        book.setCategories(new HashSet<>(
+                categoryRepository.findAllById(requestDto.getCategoryIds())));
```

Comment:
There is too much nesting and mapping done here; it's better to separate result of findAll method like so 
```suggestion
List<Categories> categories =  categoryRepository.findAllById(requestDto.getCategoryIds());
Set<Categories> uniqueCategories = categories.stream().collect(Collectors.toSet());
book.setCategories(uniqueCategories);
```